### PR TITLE
dogstatsd: implement clean disconnection for uds streams

### PR DESCRIPTION
**Pull request by iksaif**
### What does this PR do?

Implement clean disconnection for UDS Stream

### Motivation

While the current code, it closes the file descriptors abruptly when main exits. This is more gracefully and allow the listener to be stopped without stopping the entire program.
### Additional Notes

This should be reusable for TCP and named_pipe, the named_pipe refactor will have to be done by a person with Windows access.

### Possible Drawbacks / Trade-offs

 It will also minimise packet loss on stop by notifying the client of the disconnection and waiting a bit but this might not working fully depending on the entire shutdown sequence. This is also a poor implementation and a change of protocol would be necessary to avoid packet loss on stop.

### Describe how to test/QA your changes

Start/stop the agent with unix streams clients.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/dogstatsd/listeners/connections_tracker.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+			t.connections[conn] = struct{}{}
+			t.activeConnections.Inc()
+		case conn := <-t.connToClose:
+			err := conn.Close()
+			log.Infof("dogstatsd-%s: failed to close connection: %v", t.name, err)
+
+			delete(t.connections, conn)
+			t.activeConnections.Dec()
+			if requestStop && len(t.connections) == 0 {
```
Comment (vickenty) [inline comp/dogstatsd/listeners/connections_tracker.go:79]: I think this is racy and will deadlock if, assuming connections is not empty, a new connection is accepted shortly before `Stop()`, and `connToTrack` branch adds a new connection between `stopChan` and `connToClose` branches run.

Comment (iksaif) [reply comp/dogstatsd/listeners/connections_tracker.go:79]: That's true, if the contract is that we close the listening socket before then it should be fine no ?
Any other suggestion ?

**Review by vickenty — APPROVED** [review]

**Review by L3n41c — APPROVED** [review]
👍🏻 for @DataDog/container-integrations owned files.

**Comment by zdyj3170101136** [comment]
why do not gracefully close unixgram?
if close the conn directly, there would have many unread packets?

**Comment by iksaif** [comment]
- We only close the UDS datagram socket on shutdown.
- We only close UDS stream sockets on EOF
- When the client closes the unix stream socket the server can still read the buffered data.

Did you notice anything specific that could be improved ?

**Comment by zdyj3170101136** [comment]
> * We only close the UDS datagram socket on shutdown.
> * We only close UDS stream sockets on EOF
> * When the client closes the unix stream socket the server can still read the buffered data.
> 
> Did you notice anything specific that could be improved ?

if the agent want to stop, like upgrade.

it could first closeWrite of unixgram, so the client would stopped writing data.
and then it could read all of data from unixgram and close it.

**Comment by zdyj3170101136** [comment]
> * We only close the UDS datagram socket on shutdown.
> * We only close UDS stream sockets on EOF
> * When the client closes the unix stream socket the server can still read the buffered data.
> 
> Did you notice anything specific that could be improved ?

i noticed that if we stop the unixgram directly there would have many unread packets in conn.

is there any way to read it?

**Comment by zdyj3170101136** [comment]
i also wonder why we use unix stream?
is unix stream have any advantage compared to unixgram?

**Comment by zdyj3170101136** [comment]
i would be very greetful if you could help me.
@iksaif iksaif

**Pull request merged**